File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/src (BUX)/Model/Statements/AssignStmt.java
package Model.Statements;

import Exceptions.*;
import Model.*;
import Model.Expressions.Exp;
import Model.Utils.MyIDictionary;
import Model.Utils.MyIHeap;
import Model.Utils.MyIStack;

public class AssignStmt implements IStmt{
    String id;
    Exp exp;
    public AssignStmt(String id, Exp exp){
        this.id=id;
        this.exp=exp;
    }
    public String toString(){ return id+"="+exp;
    }
    public PrgState execute(PrgState state) throws MyControllerException {
        MyIStack<IStmt> stk=state.getStk();
        MyIDictionary<String,Integer> symTbl= state.getSymTable();
        MyIHeap<Integer> heap=state.getHeap();
        try {
            int val = exp.eval(symTbl,heap);
            if (symTbl.isDefined(id)==1)
                symTbl.update(id, val);
            else
                symTbl.put(id,val);
            return null;
        }
        catch (MyStmtException e){
            throw new MyControllerException(e);
        }
        catch (DivisionByZeroException e){
            throw new MyControllerException(e);
        }
    }

}

File: 3rd Semester/Advanced Programming Methods/Laboratory/Toy Interpreter C#/sharplab/Model/Statements/AssignStmt.cs
using sharplab.Model.Expressions;
using sharplab.Model.Utils;
using System;
using System.Collections.Generic;
using System.Text;

namespace sharplab.Model.Statements
{
    class AssignStmt : IStmt
    {
        string id;
        Exp exp;
        public AssignStmt(String id, Exp exp)
        {
            this.id = id;
            this.exp = exp;
        }
        public override string ToString()
        {
            return id + "=" + exp;
        }

        public PrgState Execute(PrgState state)
        {
            IMyStack<IStmt> stk=state.GetStk();
            IMyDictionary<string, int> symTbl = state.GetSymTable();
            try {
                int val = exp.Eval(symTbl);
                if (symTbl.IsDefined(id) == 1)
                    symTbl.Update(id, val);
                else
                    symTbl.Put(id, val);
                return state;
            }
            catch
            {
                throw new Exception("Invalid expression!");
            }
            
        
    }
    }
}
